src/lib/Shop.js

```diff
@@ -339,74 +360,119 @@ class AutomationShop
-                        return this.__internal__isPokeMarkUnlocked()
-                            || (TownList["Fuchsia City"].isUnlocked() && Automation.Utils.Route.canMoveToRegion(0))
-                            || TownList["Blackthorn City"].isUnlocked();
-                    }.bind(this);
-            }
-            else if ((item.name == "Lucky_egg") || (item.name == "Token_collector"))
-            {
-                itemData.isUnlocked = function() { return TownList["Pewter City"].isUnlocked(); };
-                itemData.isPurchasable = function()
+                        sellableItems[item.name] = {item: item, towns: [townName], requirements: []};
+                    }
+                    else
                     {
-                        return this.__internal__isPokeMarkUnlocked()
-                            || (TownList["Pewter City"].isUnlocked() && Automation.Utils.Route.canMoveToRegion(0))
-                            || TownList["Violet City"].isUnlocked();
-                    }.bind(this);
-            }
-            else if ((item.name == "Dowsing_machine") || (item.name == "Lucky_incense"))
-            {
-                itemData.isUnlocked = function() { return TownList["Lavender Town"].isUnlocked(); };
-                itemData.isPurchasable = function()
+                        sellableItems[item.name].towns.push(townName);
+                    }
+
+                    // Add the requirements
+                    if (shop.requirements !== undefined && shop.requirements.length > 0)
                     {
-                        return this.__internal__isPokeMarkUnlocked()
-                            || (TownList["Lavender Town"].isUnlocked() && Automation.Utils.Route.canMoveToRegion(0))
-                            || TownList["Olivine City"].isUnlocked();
-                    }.bind(this);
-            }
-            else if (item.name == "SmallRestore")
-            {
-                itemData.isUnlocked = function() { return TownList["Cinnabar Island"].isUnlocked(); };
-                itemData.isPurchasable = itemData.isUnlocked;
+                        sellableItems[item.name].requirements.push(shop.requirements);
+                    }
+                    else
+                    {
+                        sellableItems[item.name].requirements.push([]);
+                    }
+                }
+
             }
-            else if (item.name == "MediumRestore")
+        }
+
+        // Iterate with of instead of Object.entries to keep the order
+        for (const itemData of Object.values(sellableItems))
+        {
+            let shopItem = { item: itemData.item }
+            shopItem.isUnlocked = function() {
+                return itemData.towns.some(townName => TownList[townName].isUnlocked())
+                    && (itemData.requirements.length == 0
+                        || itemData.requirements.some(requirementList => requirementList.every(requirement => requirement.isCompleted())));
```

Comment:
Do we really need to manage the town requirement ?
I feel like if the town is unlocked, then all requirements are met

https://github.com/pokeclicker/pokeclicker/blob/develop/src/scripts/towns/Town.ts#L56-L58

Reply:
Technically, the only shops with requirements sell items that we dont care about (pokemons mainly), so we could ignore it.

But there are shops that are unlocked way after the town itself is unlocked, like the one that sells Zarude Dada

If some item gets added like that that we want to auto-buy, we can add the check for the requirements then.